crates/rattler_installs_packages/src/python_env/byte_code_compiler.rs

```
@@ -0,0 +1,279 @@
+        // to exit.
+        drop(self.request_tx.take());
+
+        // Wait for the compilation host to exit
+        self.child
+            .take()
+            .expect("the child is only dropped on drop")
+            .wait()?;
+
+        Ok(())
+    }
+}
+
+impl Drop for ByteCodeCompiler {
+    fn drop(&mut self) {
+        drop(self.request_tx.take());
+        if let Some(mut child) = self.child.take() {
+            child.wait().unwrap();
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::python_env::system_python_executable;
+
+    #[test]
+    fn test() {
+        let python_path = system_python_executable().unwrap();
+
+        // Create a temporary file that holds the sourcecode of the compilation host
+        let mut compiler_source = tempfile::Builder::new().suffix(".py").tempfile().unwrap();
+        compiler_source
+            .write_all(include_bytes!("compile_pyc.py"))
+            .unwrap();
+
+        // Spawn a compiler and compile the compilation host source code.
+        let compiler = ByteCodeCompiler::new(&python_path).unwrap();
+        let pyc_file = compiler.compile_and_wait(compiler_source.path()).unwrap();
+
+        // Make sure the compiled file exists
+        assert!(pyc_file.is_file(), "The compiled file does not exist!");
+
+        // Wait should immediately return
+        compiler.wait().unwrap();
+    }
+
+    #[test]
+    fn test_failed_case() {
+        let python_path = system_python_executable().unwrap();
+
+        // Create a temporary file that holds the sourcecode of the compilation host
+        let mut compiler_source = tempfile::Builder::new().suffix(".py").tempfile().unwrap();
+        compiler_source.write_all(b"$").unwrap();
+
+        // Spawn a compiler and compile the compilation host source code.
+        let compiler = ByteCodeCompiler::new(&python_path).unwrap();
+        compiler
+            .compile_and_wait(compiler_source.path())
+            .unwrap_err();
```

Does this mean the process will fail when it can't compile a file? I think usually we should ignore failures – it can happen that some Python file can only be compiled with a newer version, or has some template variables in it etc. so we should silently fail (or with a trace::warn or so).

---

Yes! I will change the unpack behavior to just ignore these errors.